### PR TITLE
Generics

### PR DESCRIPTION
**Pull request by certik**
This is a long running PR that implements generics in LFortran.

### CODE REVIEW

**Comment by certik** [comment]
TODO:
* [ ] Get the `examples/swap_m.f90` compiling to ASR. Add an ASR test.
* [ ] We'll add code that calls `swap` and we will instantiate the function. For the actual instantiation, we will:
  * [ ] Port LPython's generics to libasr and LFortran (https://github.com/lcompilers/lpython/pull/625)

File: src/libasr/ASR.asdl [hunk 1]
```diff
@@ -87,6 +87,19 @@ symbol
     | Subroutine(symbol_table symtab, identifier name, expr* args, stmt* body,
         abi abi, access access, deftype deftype, string? bindc_name, bool pure,
         bool module)
+
+-- restrictions is a list of Restriction
+-- fn is a list of TempaltedFunction
```
Comment (namannimmo10) [inline src/libasr/ASR.asdl:92]:
```suggestion
-- fn is a list of TemplatedFunction
```

**Comment by certik** [comment]
Short term TODO:

* [ ] We work in `src/lfortran/semantics/ast_symboltable_visitor.cpp`. Implement `visit_Template()` and ignore everything, just visit every function in the `m_contains` member.
* [ ] In `visit_Function` in the symbol table visitor, we check if any argument is of type `TypeParameter`, if so, it is a generic. In that case, return `TemplatedFunction` instead of a `Function`.

**Comment by certik** [comment]
Closing in favor of #39.

**Pull request closed without merging**